File: java/src/main/java/org/example/auth/AWS4SignerBase.java
package org.example.auth;

import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.SimpleTimeZone;
import java.util.TreeMap;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import org.example.Utility.NetUtils;

import com.amazonaws.util.BinaryUtils;

public abstract class AWS4SignerBase {

	public static final String EMPTY_BODY_SHA256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";
	public static final String UNSIGNED_PAYLOAD = "UNSIGNED-PAYLOAD";

	public static final String SCHEME = "AWS4";
	public static final String ALGORITHM = "HMAC-SHA256";
	public static final String TERMINATOR = "aws4Request";
	public static final String X_AMZ_CONTENT_SHA256 = "X-Amz-Content-SHA256";

	public static final String ISO8601_BASIC_FORMAT = "yyyyMMdd'T'HHmmss'Z'";
	public static final String DATE_STRING_FORMAT = "yyyyMMdd";

	protected URL endpointUrl;
	protected String httpMethod;
	protected String serviceName;
	protected String regionName;

	protected final SimpleDateFormat dateTimeFormat;
	protected final SimpleDateFormat dateStampFormat;

	public AWS4SignerBase(URL endpointUrl, String httpMethod, String serviceName, String regionName) {
		this.endpointUrl = endpointUrl;
		this.httpMethod = httpMethod;
		this.serviceName = serviceName;
		this.regionName = regionName;

		dateTimeFormat = new SimpleDateFormat(ISO8601_BASIC_FORMAT);
		dateTimeFormat.setTimeZone(new SimpleTimeZone(0, "UTC"));
		dateStampFormat = new SimpleDateFormat(DATE_STRING_FORMAT);
		dateStampFormat.setTimeZone(new SimpleTimeZone(0, "UTC"));
	}

	protected static String getCanonicalizeHeaderNames(Map<String, String> headers) {
		List<String> sortedHeaders = new ArrayList<>();
		sortedHeaders.addAll(headers.keySet());
		Collections.sort(sortedHeaders, String.CASE_INSENSITIVE_ORDER);

		StringBuilder buffer = new StringBuilder();
		for (String header : sortedHeaders) {
			if (buffer.length() > 0)
				buffer.append(";");
			buffer.append(header.toLowerCase());
		}

		return buffer.toString();
	}

	/**
	 * Computes the canonical headers with values for the request. For AWS4, all
	 * headers must be included in the signing process.
	 */
	protected static String getCanonicalizedHeaderString(Map<String, String> headers) {
		if (headers == null || headers.isEmpty()) {
			return "";
		}

		// step1: sort the headers by case-insensitive order
		List<String> sortedHeaders = new ArrayList<>();
		sortedHeaders.addAll(headers.keySet());
		Collections.sort(sortedHeaders, String.CASE_INSENSITIVE_ORDER);

		// step2: form the canonical header:value entries in sorted order.
		// Multiple white spaces in the values should be compressed to a single
		// space.
		StringBuilder buffer = new StringBuilder();
		for (String key : sortedHeaders) {
			buffer.append(key.toLowerCase().replaceAll("\\s+", " ") + ":" + headers.get(key).replaceAll("\\s+", " "));
			buffer.append("\n");
		}

		return buffer.toString();
	}

	/**
	 * Returns the canonical request string to go into the signer process; this
	 * consists of several canonical sub-parts.
	 * 
	 * @return
	 */
	protected static String getCanonicalRequest(URL endpoint, String httpMethod, String queryParameters,
			String canonicalizedHeaderNames, String canonicalizedHeaders, String bodyHash) {
		return httpMethod + "\n" +
				getCanonicalizedResourcePath(endpoint) + "\n" +
				queryParameters + "\n" +
				canonicalizedHeaders + "\n" +
				canonicalizedHeaderNames + "\n" +
				bodyHash;
	}

	/**
	 * Returns the canonicalized resource path for the service endpoint.
	 */
	protected static String getCanonicalizedResourcePath(URL endpoint) {
		if (endpoint == null) {
			return "/";
		}
		String path = endpoint.getPath();
		if (path == null || path.isEmpty()) {
			return "/";
		}

		String encodedPath = NetUtils.urlEncode(path, true);
		if (encodedPath.startsWith("/")) {
			return encodedPath;
		} else {
			return "/".concat(encodedPath);
		}
	}

	/**
	 * Examines the specified query string parameters and returns a
	 * canonicalized form.
	 * <p>
	 * The canonicalized query string is formed by first sorting all the query
	 * string parameters, then URI encoding both the key and value and then
	 * joining them, in order, separating key value pairs with an '&'.
	 *
	 * @param parameters
	 *                   The query string parameters to be canonicalized.
	 *
	 * @return A canonicalized form for the specified query string parameters.
	 */
	public static String getCanonicalizedQueryString(Map<String, String> parameters) {
		if (parameters == null || parameters.isEmpty()) {
			return "";
		}

		var sorted = new TreeMap<String, String>();

		var pairs = parameters.entrySet().iterator();
		while (pairs.hasNext()) {
			var pair = pairs.next();
			var key = pair.getKey();
			var value = pair.getValue();
			sorted.put(NetUtils.urlEncode(key, false), NetUtils.urlEncode(value, false));
		}

		StringBuilder builder = new StringBuilder();
		pairs = sorted.entrySet().iterator();
		while (pairs.hasNext()) {
			var pair = pairs.next();
			builder.append(pair.getKey());
			builder.append("=");
			builder.append(pair.getValue());
			if (pairs.hasNext()) {
				builder.append("&");
			}
		}

		return builder.toString();
	}

	protected static String getStringToSign(String scheme, String algorithm, String dateTime, String scope,
			String canonicalRequest) {
		return scheme + "-" + algorithm + "\n" +
				dateTime + "\n" +
				scope + "\n" +
				BinaryUtils.toHex(hash(canonicalRequest));
	}

	public static byte[] hash(String text) {
		try {
			MessageDigest md = MessageDigest.getInstance("SHA-256");
			md.update(text.getBytes(StandardCharsets.UTF_8));
			return md.digest();
		} catch (Exception e) {
			throw new RuntimeException("Unable to compute hash while signing request: " + e.getMessage(), e);
		}
	}

	public static byte[] hash(byte[] data) {
		try {
			MessageDigest md = MessageDigest.getInstance("SHA-256");
			md.update(data);
			return md.digest();
		} catch (Exception e) {
			throw new RuntimeException("Unable to compute hash while signing request: " + e.getMessage(), e);
		}
	}

	protected static byte[] sign(String stringData, byte[] key, String algorithm) {
		try {
			byte[] data = stringData.getBytes(StandardCharsets.UTF_8);
			Mac mac = Mac.getInstance(algorithm);
			mac.init(new SecretKeySpec(key, algorithm));
			return mac.doFinal(data);
		} catch (Exception e) {
			throw new RuntimeException("Unable to calculate a request signature: " + e.getMessage(), e);
		}
	}
}

File: python/s3tests/auth/aws4_signer_base.py
"""AWS Signature Version 4 base helpers ported from Java AWS4SignerBase."""

from __future__ import annotations

import hashlib
import hmac
from datetime import datetime, timezone
from typing import Dict, Optional
from urllib.parse import quote, urlparse

from s3tests.utils import net_utils

EMPTY_BODY_SHA256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
UNSIGNED_PAYLOAD = "UNSIGNED-PAYLOAD"
SCHEME = "AWS4"
ALGORITHM = "HMAC-SHA256"
TERMINATOR = "aws4_request"
X_AMZ_CONTENT_SHA256 = "X-Amz-Content-SHA256"
ISO8601_BASIC_FORMAT = "%Y%m%dT%H%M%SZ"
DATE_STRING_FORMAT = "%Y%m%d"


def get_amz_date(dt: Optional[datetime] = None) -> str:
    value = dt or datetime.now(timezone.utc)
    return value.strftime(ISO8601_BASIC_FORMAT)


def get_post_policy_signature(secret_key: str, date_stamp: str, region_name: str, policy_base64: str) -> str:
    k_secret = (SCHEME + secret_key).encode("utf-8")
    k_date = _sign(date_stamp, k_secret)
    k_region = _sign(region_name, k_date)
    k_service = _sign("s3", k_region)
    k_signing = _sign(TERMINATOR, k_service)
    return _to_hex(_sign(policy_base64, k_signing))


def hash_text(text: str) -> bytes:
    return hashlib.sha256(text.encode("utf-8")).digest()


def hash_bytes(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def _to_hex(data: bytes) -> str:
    return data.hex()


def _sign(data: str, key: bytes) -> bytes:
    return hmac.new(key, data.encode("utf-8"), hashlib.sha256).digest()


def canonicalize_header_names(headers: Dict[str, str]) -> str:
    return ";".join(sorted((k.lower() for k in headers.keys()), key=str.lower))


def canonicalized_header_string(headers: Dict[str, str]) -> str:
    if not headers:
        return ""
    lines = []
    for key in sorted(headers.keys(), key=str.lower):
        value = " ".join(headers[key].split())
        lines.append(f"{key.lower()}:{value}")
    return "\n".join(lines) + "\n"


def canonicalized_resource_path(endpoint: str) -> str:
    path = urlparse(endpoint).path or "/"
    encoded = net_utils.url_encode(path, keep_path_slash=True)
    return encoded if encoded.startswith("/") else f"/{encoded}"


def format_host_header(endpoint: str) -> str:
    parsed = urlparse(endpoint)
    host = parsed.hostname or ""
    if parsed.port and parsed.port not in (80, 443):
        return f"{host}:{parsed.port}"
    return host


def canonicalized_query_string(parameters: Optional[Dict[str, str]]) -> str:
    if not parameters:
        return ""
    items = []
    for key in sorted(parameters.keys()):
        items.append(
            f"{net_utils.url_encode(key, keep_path_slash=False)}="
            f"{net_utils.url_encode(parameters[key], keep_path_slash=False)}"
        )
    return "&".join(items)


def canonical_request(
    endpoint: str,
    http_method: str,
    query_parameters: str,
    canonicalized_header_names: str,
    canonicalized_headers: str,
    body_hash: str,
) -> str:
    return "\n".join(
        [
            http_method,
            canonicalized_resource_path(endpoint),
            query_parameters,
            canonicalized_headers,
            canonicalized_header_names,
            body_hash,
        ]
    )


def string_to_sign(scheme: str, algorithm: str, date_time: str, scope: str, canonical_request_text: str) -> str:
    return "\n".join(
        [
            f"{scheme}-{algorithm}",
            date_time,
            scope,
            _to_hex(hash_text(canonical_request_text)),
        ]
    )


def signing_key(secret_key: str, date_stamp: str, region_name: str, service_name: str) -> bytes:
    k_secret = (SCHEME + secret_key).encode("utf-8")
    k_date = _sign(date_stamp, k_secret)
    k_region = _sign(region_name, k_date)
    k_service = _sign(service_name, k_region)
    return _sign(TERMINATOR, k_service)
